### ISSUE TITLE
"rustc does not exist" when cross-compiling ARM on Windows

### ISSUE THREAD

**Issue by Timmmm** [rust-lang-nursery/rustup.rs#672]
On Windows 7, I installed rustup, and then did:

    rustup install stable-armv7-unknown-linux-gnueabihf

(Or something like that.) Unfortunately it didn't seem to install rustc, or can't find it:

```
>rustup show
Default host: x86_64-pc-windows-msvc

installed toolchains
--------------------

stable-armv7-unknown-linux-gnueabihf (default)
stable-x86_64-pc-windows-msvc

active toolchain
----------------

stable-armv7-unknown-linux-gnueabihf (default)
(rustc does not exist)

```

If I try to run `cargo`:

```
>cargo
error: command failed: 'cargo'
info: caused by: %1 is not a valid Win32 application. (os error 193)
```

**Comment by Timmmm** [comment]
In fact, there *is* a `rustc.exe`, in `~/.cargo/bin`. Windows just doesn't think it is a valid application.

I opened it in depends.exe and it seems there is an architecture problem. `rustc.exe` is an x86 binary, but all the DLLs it tries to load are x64 (I'm on 64-bit Windows).

**Comment by Diggsey** [comment]
This is not how you cross-compile with rustup. This will install the ARM toolchain onto your computer (which will not be a valid win32 application!)

See this blog post for more information, it has examples of cross compiling for musl and android: https://blog.rust-lang.org/2016/05/13/rustup.html

TLDR: You want to use a *toolchain* compatible with your *host* OS (windows), and install the ARM *target* within that toolchain.

You will run into difficulties, because you will need an ELF linker which can run on windows (which is just barely possible: see the GOLD linker which comes with mingw) and you'll also need any system libraries that you're going to link against (can't help you here I'm afraid).

I've managed to cross-compile from windows to linux-musl previously, which obviates the need for any additional system libraries to be available, and using the GOLD linker, but it wasn't compiling for ARM.

I am working on a project to make this easier, but cross-compiling from windows to non-musl linux is always going to be problematic.

**Comment by Timmmm** [comment]
Ah I see, my mistake. I guess I'll have to wait for an `armv7-unknown-linux-muscl` target. Somewhat disappointing that this isn't as easy as Go.

**Comment by Diggsey** [comment]
Go programs make system calls directly, do not depend on any shared libraries, and they even have their own linker.

For pure Go programs, this makes cross-compilation much easier, but if you use *any* C code or dependencies in your program, you're in trouble.
